cpp/include/raft/sparse/solver/detail/lanczos.cuh

```diff
@@ -1396,4 +1435,699 @@ int computeLargestEigenvectors(
+                             return y / *device_scalar;
+                           });
+
+    lanczos_aux(handle,
+                A,
+                V.view(),
+                u.view(),
+                alpha.view(),
+                beta.view(),
+                nEigVecs + 1,
+                ncv,
+                ncv,
+                v.view(),
+                aux_uu.view(),
+                vv.view());
+    iter += ncv - nEigVecs;
+    lanczos_solve_ritz<IndexTypeT, ValueTypeT>(handle,
+                                               alpha.view(),
+                                               beta.view(),
+                                               beta_k.view(),
+                                               nEigVecs,
+                                               0,
+                                               ncv,
+                                               eigenvectors.view(),
+                                               eigenvalues.view());
+    auto eigenvectors_k = raft::make_device_matrix_view<ValueTypeT, uint32_t, raft::col_major>(
+      eigenvectors.data_handle(), ncv, nEigVecs);
+
+    auto ritz_eigenvectors = raft::make_device_matrix_view<ValueTypeT, uint32_t, raft::col_major>(
+      eigVecs_dev, n, nEigVecs);
+
+    auto V_T =
+      raft::make_device_matrix_view<ValueTypeT, uint32_t, raft::col_major>(V.data_handle(), n, ncv);
+    raft::linalg::gemm<ValueTypeT, uint32_t, raft::col_major, raft::col_major, raft::col_major>(
+      handle, V_T, eigenvectors_k, ritz_eigenvectors);
+
+    auto eigenvectors_k_slice =
+      raft::make_device_matrix_view<ValueTypeT, IndexTypeT, raft::col_major>(
+        eigenvectors.data_handle(), ncv, nEigVecs);
+    auto S_matrix = raft::make_device_matrix_view<ValueTypeT, IndexTypeT, raft::col_major>(
+      s.data_handle(), 1, nEigVecs);
+
+    raft::matrix::slice_coordinates<IndexTypeT> coords(ncv - 1, 0, ncv, nEigVecs);
+    raft::matrix::slice(handle, make_const_mdspan(eigenvectors_k_slice), S_matrix, coords);
+
+    raft::matrix::fill(handle, beta_k.view(), zero);
+
+    auto beta_scalar =
+      raft::make_device_scalar_view<const ValueTypeT>(&((beta.view())(0, ncv - 1)));
+
+    raft::linalg::axpy(handle, beta_scalar, raft::make_const_mdspan(s.view()), beta_k.view());
+
+    raft::device_vector<ValueTypeT, uint32_t> output2 =
+      raft::make_device_vector<ValueTypeT, uint32_t>(handle, 1);
+    raft::device_matrix_view<const ValueTypeT> input2 =
+      raft::make_device_matrix_view<const ValueTypeT>(beta_k.data_handle(), 1, nEigVecs);
+    raft::linalg::norm(handle,
+                       input2,
+                       output2.view(),
+                       raft::linalg::L2Norm,
+                       raft::linalg::Apply::ALONG_ROWS,
+                       raft::sqrt_op());
+    raft::copy(&res, output2.data_handle(), 1, stream);
+
+    RAFT_LOG_TRACE("Iteration %f: residual (tolerance) %d", iter, res);
```

Comment:
To use a copied value on host, the stream would need to be synchronized before. Since that sync can slow this function down it would be better to check the log level, and only copy and sync if necessary.